cpmpy/transformations/normalize.py

```diff
@@ -55,35 +56,41 @@ def simplify_boolean(lst_of_expr, num_context=False):
     newlist = []
     for expr in lst_of_expr:
 
-        if isinstance(expr, bool):
-            # not sure if this should happen here or at construction time
-            expr = BoolVal(expr)
-
-        if isinstance(expr, BoolVal):
-            newlist.append(int(expr.value()) if num_context else expr)
-
-        elif isinstance(expr, Operator):
-            args = simplify_boolean(expr.args, num_context=not expr.is_bool())
+        if isinstance(expr, Operator):
+            if expr.has_subexpr():
+                args = simplify_boolean(expr.args, num_context=not expr.is_bool())
+            else:
+                args = list(expr.args)
 
             if expr.name == "or":
-                if any(is_true_cst(arg) for arg in args):
-                    newlist.append(1 if num_context else BoolVal(True))
-                else:
-                    filtered_args = [arg for arg in args if not isinstance(arg, BoolVal)]
-                    if len(filtered_args):
-                        newlist.append(Operator("or", filtered_args))
-                    else:
+                for i,a in enumerate(list(args)):
+                    if is_true_cst(a):
+                        newlist.append(1 if num_context else BoolVal(True))
+                        break
+                    elif is_false_cst(a):
+                        args.pop(i)
```

Comment:
I think we want to do copy on write, so maybe it's better not to change the args in place here. Haven't really thought through the entire copy on write thing, but we might need to enforce that to be able to cash hashes and stuff ;)